source/shared/core_sqlsrv.h

```diff
@@ -2416,4 +2430,39 @@ struct column_encryption_set_func {
     }
 };
 
+struct ce_ksp_provider_set_func {
+
+    static void func( _In_ connection_option const* option, _In_ zval* value, _Inout_ sqlsrv_conn* conn, _Inout_ std::string& conn_str TSRMLS_DC)
+    {
+        CHECK_CUSTOM_ERROR( Z_TYPE_P( value ) != IS_STRING, conn, SQLSRV_ERROR_INVALID_OPTION_TYPE_STRING, option->odbc_name ) {
+            throw core::CoreException();
+        }
+            
+        size_t value_len = Z_STRLEN_P( value );
+
+        CHECK_CUSTOM_ERROR( value_len == 0, conn, SQLSRV_ERROR_KEYSTORE_INVALID_VALUE )  {
+            throw core::CoreException();
+        }
+
+        switch ( option->conn_option_key ) {
+            case SQLSRV_CONN_OPTION_CEKEYSTORE_PROVIDER: 
+                conn->ce_option.ksp_path = value;
+                conn->ce_option.ksp_required = true;
+                break;
+            case SQLSRV_CONN_OPTION_CEKEYSTORE_NAME:
+                conn->ce_option.ksp_name = value;
+                conn->ce_option.ksp_required = true;
+                break;
+            case SQLSRV_CONN_OPTION_CEKEYSTORE_ENCRYPT_KEY:
+                conn->ce_option.ksp_encrypt_key = value;
+                conn->ce_option.key_size = value_len;
+                conn->ce_option.ksp_required = true;
+                break;
+            default:
+                SQLSRV_ASSERT( false, "ce_ksp_provider_set_func: Invalid KSP option!" );
+                break;
+        }
+    }
+};
```

Comment:
Are all three keywords required if the user wants to use a custom keystore?

Reply:
yes, as you can see in the helper function load_configure_ksp()